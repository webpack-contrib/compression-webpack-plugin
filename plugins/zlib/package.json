{
  "name": "compression-webpack-plugin-zlib",
  "version": "1.0.0",
  "author": "Tobias Koppers @sokra",
  "maintainers": [
    "Jason Palmer @jsonp"
  ],
  "description": "Prepare compressed versions of assets to serve them with Content-Encoding",
  "dependencies": {
    "async": "0.2.x",
    "webpack-sources": "^0.1.0"
  },
  "peerDependencies": {
    "webpack": ">=0.11 <2.0.0 || >=2.1.0-beta.13 <3.0.0"
  },
  "devDependencies": {
    "mocha": "^2.3.2"
  },
  "scripts": {
    "test": "mocha --reporter min 'test/**/*test*.js'"
  },
  "files": [
    "index.js"
  ],
  "homepage": "http://github.com/webpack/compression-webpack-plugin",
  "repository": {
    "type": "git",
    "url": "http://github.com/webpack/compression-webpack-plugin.git"
  },
  "licenses": [
    {
      "type": "MIT",
      "url": "http://www.opensource.org/licenses/mit-license.php"
    }
  ]
}
