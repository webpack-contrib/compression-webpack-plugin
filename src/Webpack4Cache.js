import os from 'os';

import cacache from 'cacache';
import findCacheDir from 'find-cache-dir';
import serialize from 'serialize-javascript';

export default class Webpack4Cache {
  constructor(compilation, options, weakCache) {
    this.cache =
      options.cache === true
        ? Webpack4Cache.getCacheDirectory()
        : options.cache;
    this.weakCache = weakCache;
  }

  static getCacheDirectory() {
    return findCacheDir({ name: 'compression-webpack-plugin' }) || os.tmpdir();
  }

  async get(task, sources) {
    const weakOutput = this.weakCache.get(task.assetSource);

    if (weakOutput) {
      return weakOutput;
    }

    if (!this.cache) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    // eslint-disable-next-line no-param-reassign
    task.cacheIdent = task.cacheIdent || serialize(task.cacheKeys);

    let cachedResult;

    try {
      cachedResult = await cacache.get(this.cache, task.cacheIdent);
    } catch (ignoreError) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    return new sources.RawSource(
      Buffer.from(JSON.parse(cachedResult.data).data)
    );
  }

  async store(task) {
    if (!this.weakCache.has(task.assetSource)) {
      this.weakCache.set(task.assetSource, task.output);
    }

    if (!this.cache) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    const { cacheIdent, output } = task;

    return cacache.put(this.cache, cacheIdent, JSON.stringify(output.source()));
  }
}
