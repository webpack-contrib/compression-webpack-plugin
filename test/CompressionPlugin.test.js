import zlib from 'zlib';
import path from 'path';

import webpack from 'webpack';
import findCacheDir from 'find-cache-dir';
import cacache from 'cacache';

import Webpack4Cache from '../src/Webpack4Cache';
import CompressionPlugin from '../src/index';

import {
  compile,
  CopyPluginWithAssetInfo,
  ModifyExistingAsset,
  getAssetsNameAndSize,
  getCompiler,
  getErrors,
  getWarnings,
  readAsset,
  removeCache,
} from './helpers/index';

const cacheDir1 = findCacheDir({ name: 'compression-webpack-plugin-cache-1' });
const cacheDir2 = findCacheDir({ name: 'compression-webpack-plugin-cache-2' });
const cacheDir3 = findCacheDir({ name: 'compression-webpack-plugin-cache-3' });

describe('CompressionPlugin', () => {
  beforeAll(() => {
    return Promise.all([
      removeCache(),
      cacache.rm.all(cacheDir1),
      cacache.rm.all(cacheDir2),
      cacache.rm.all(cacheDir3),
    ]);
  });

  it('should work', async () => {
    expect.assertions(6);

    const compiler = getCompiler(
      './entry.js',
      {},
      {
        output: {
          path: `${__dirname}/dist`,
          filename: '[name].js?var=[hash]',
          chunkFilename: '[id].[name].js?ver=[hash]',
        },
      }
    );

    new CompressionPlugin().apply(compiler);

    const stats = await compile(compiler);
    const { assets, assetsInfo } = stats.compilation;

    for (const assetName of Object.keys(assets)) {
      const info = assetsInfo.get(assetName);

      if (!info.related) {
        // eslint-disable-next-line no-continue
        continue;
      }

      const originalBuffer = readAsset(assetName, compiler, stats);
      const gzipedBuffer = readAsset(info.related.gziped, compiler, stats);

      expect(zlib.gunzipSync(gzipedBuffer).equals(originalBuffer)).toBe(true);
    }

    expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');
  });

  it('should work with assets info', async () => {
    const compiler = getCompiler(
      './entry.js',
      {},
      {
        devtool: 'source-map',
        output: {
          path: `${__dirname}/dist`,
          filename: '[name].js?var=[hash]',
          chunkFilename: '[id].[name].js?ver=[hash]',
        },
      }
    );

    new CompressionPlugin().apply(compiler);
    new CopyPluginWithAssetInfo().apply(compiler);

    const stats = await compile(compiler);

    expect(getAssetsNameAndSize(stats, true)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');
  });

  it('should work child compilations', async () => {
    const gzipSpy = jest.spyOn(zlib, 'gzip');
    const compiler = getCompiler(
      './entry.js',
      {},
      {
        output: {
          path: `${__dirname}/dist`,
          filename: '[name].js?var=[hash]',
          chunkFilename: '[id].[name].js?ver=[hash]',
        },
        module: {
          rules: [
            {
              test: /number\.js$/i,
              rules: [
                {
                  loader: require.resolve(
                    './helpers/loader-with-child-compilation.js'
                  ),
                },
              ],
            },
            {
              test: /\.(png|jpg|gif|svg)$/i,
              rules: [
                {
                  loader: 'file-loader',
                },
              ],
            },
          ],
        },
      }
    );

    new CompressionPlugin().apply(compiler);

    const stats = await compile(compiler);

    expect(gzipSpy).toHaveBeenCalledTimes(5);
    expect(getAssetsNameAndSize(stats, true)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');

    gzipSpy.mockRestore();
  });

  it('should work with multiple plugins', async () => {
    const compiler = getCompiler(
      './entry.js',
      {},
      {
        output: {
          path: `${__dirname}/dist`,
          filename: '[name].js?var=[hash]',
          chunkFilename: '[id].[name].js?ver=[hash]',
        },
      }
    );

    new CompressionPlugin({
      algorithm: 'gzip',
      filename: '[path][base].gz',
    }).apply(compiler);
    new CompressionPlugin({
      algorithm: 'brotliCompress',
      filename: '[path][base].br',
    }).apply(compiler);
    new CompressionPlugin({
      algorithm: (input, options, callback) => {
        return callback(input);
      },
      filename: '[path][base].compress',
    }).apply(compiler);
    new CompressionPlugin({
      algorithm: (input, options, callback) => {
        return callback(input);
      },
      filename: '[path][base].custom?foo=bar#hash',
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(getAssetsNameAndSize(stats, true)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');
  });

  it('should work and show compress assets in stats', async () => {
    const compiler = getCompiler(
      './entry.js',
      {},
      {
        stats: 'verbose',
        output: {
          path: `${__dirname}/dist`,
          filename: '[name].js',
          chunkFilename: '[id].[name].js',
        },
      }
    );

    new CompressionPlugin().apply(compiler);

    const stats = await compile(compiler);
    const stringStats = stats.toString({ relatedAssets: true });
    const printedCompressed = stringStats.match(/\[compressed]/g);

    expect(printedCompressed ? printedCompressed.length : 0).toBe(
      getCompiler.isWebpack4() ? 0 : 3
    );
    expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');
  });

  it('should work and keep assets info', async () => {
    const compiler = getCompiler(
      './entry.js',
      {},
      {
        stats: 'verbose',
        output: {
          path: `${__dirname}/dist`,
          filename: '[name].[contenthash].js',
          chunkFilename: '[id].[name].[contenthash].js',
        },
      }
    );

    new CompressionPlugin().apply(compiler);

    const stats = await compile(compiler);

    for (const [, info] of stats.compilation.assetsInfo.entries()) {
      expect(info.immutable).toBe(true);
    }

    expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');
  });

  it('should work and use memory cache without options in the "development" mode', async () => {
    const getCacheDirectorySpy = jest
      .spyOn(Webpack4Cache, 'getCacheDirectory')
      .mockImplementation(() => {
        return cacheDir1;
      });

    const compiler = getCompiler('./entry.js', {}, { mode: 'development' });

    new CompressionPlugin().apply(compiler);

    const stats = await compile(compiler);

    if (webpack.version[0] === '4') {
      expect(
        Object.keys(stats.compilation.assets).filter(
          (assetName) => stats.compilation.assets[assetName].emitted
        ).length
      ).toBe(7);
    } else {
      expect(stats.compilation.emittedAssets.size).toBe(7);
    }

    expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');

    await new Promise(async (resolve) => {
      const newStats = await compile(compiler);

      if (webpack.version[0] === '4') {
        expect(
          Object.keys(newStats.compilation.assets).filter(
            (assetName) => newStats.compilation.assets[assetName].emitted
          ).length
        ).toBe(0);
      } else {
        expect(newStats.compilation.emittedAssets.size).toBe(0);
      }

      expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
      expect(getWarnings(newStats)).toMatchSnapshot('errors');
      expect(getErrors(newStats)).toMatchSnapshot('warnings');

      getCacheDirectorySpy.mockRestore();

      resolve();
    });
  });

  it('should work and use memory cache when the "cache" option is "true"', async () => {
    const getCacheDirectorySpy = jest
      .spyOn(Webpack4Cache, 'getCacheDirectory')
      .mockImplementation(() => {
        return cacheDir2;
      });

    const compiler = getCompiler(
      './entry.js',
      {},
      {
        cache: true,
        output: {
          path: path.resolve(__dirname, './outputs'),
          filename: '[name].js',
          chunkFilename: '[id].js',
        },
      }
    );

    new CompressionPlugin().apply(compiler);

    const stats = await compile(compiler);

    if (webpack.version[0] === '4') {
      expect(
        Object.keys(stats.compilation.assets).filter(
          (assetName) => stats.compilation.assets[assetName].emitted
        ).length
      ).toBe(7);
    } else {
      expect(stats.compilation.emittedAssets.size).toBe(7);
    }

    expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');

    await new Promise(async (resolve) => {
      const newStats = await compile(compiler);

      if (webpack.version[0] === '4') {
        expect(
          Object.keys(newStats.compilation.assets).filter(
            (assetName) => newStats.compilation.assets[assetName].emitted
          ).length
        ).toBe(0);
      } else {
        expect(newStats.compilation.emittedAssets.size).toBe(0);
      }

      expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
      expect(getWarnings(newStats)).toMatchSnapshot('errors');
      expect(getErrors(newStats)).toMatchSnapshot('warnings');

      getCacheDirectorySpy.mockRestore();

      resolve();
    });
  });

  it('should work and use memory cache when the "cache" option is "true" and the asset has been changed', async () => {
    const getCacheDirectorySpy = jest
      .spyOn(Webpack4Cache, 'getCacheDirectory')
      .mockImplementation(() => {
        return cacheDir3;
      });

    const compiler = getCompiler(
      './entry.js',
      {},
      {
        cache: true,
        output: {
          path: path.resolve(__dirname, './outputs'),
          filename: '[name].js',
          chunkFilename: '[id].js',
        },
      }
    );

    new CompressionPlugin().apply(compiler);

    const stats = await compile(compiler);

    if (webpack.version[0] === '4') {
      expect(
        Object.keys(stats.compilation.assets).filter(
          (assetName) => stats.compilation.assets[assetName].emitted
        ).length
      ).toBe(7);
    } else {
      expect(stats.compilation.emittedAssets.size).toBe(7);
    }

    expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');

    new ModifyExistingAsset({ name: 'main.js' }).apply(compiler);

    await new Promise(async (resolve) => {
      const newStats = await compile(compiler);

      if (webpack.version[0] === '4') {
        expect(
          Object.keys(newStats.compilation.assets).filter(
            (assetName) => newStats.compilation.assets[assetName].emitted
          ).length
        ).toBe(2);
      } else {
        expect(newStats.compilation.emittedAssets.size).toBe(2);
      }

      expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
      expect(getWarnings(newStats)).toMatchSnapshot('errors');
      expect(getErrors(newStats)).toMatchSnapshot('warnings');

      getCacheDirectorySpy.mockRestore();

      resolve();
    });
  });

  it('should work and do not use memory cache when the "cache" option is "false"', async () => {
    const compiler = getCompiler(
      './entry.js',
      {
        name: '[name].[ext]',
      },
      {
        cache: false,
        output: {
          path: path.resolve(__dirname, './outputs'),
          filename: '[name].js',
          chunkFilename: '[id].[name].js',
        },
      }
    );

    new CompressionPlugin(
      webpack.version[0] === '4' ? { cache: false } : {}
    ).apply(compiler);

    const stats = await compile(compiler);

    if (webpack.version[0] === '4') {
      expect(
        Object.keys(stats.compilation.assets).filter(
          (assetName) => stats.compilation.assets[assetName].emitted
        ).length
      ).toBe(7);
    } else {
      expect(stats.compilation.emittedAssets.size).toBe(7);
    }

    expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
    expect(getWarnings(stats)).toMatchSnapshot('errors');
    expect(getErrors(stats)).toMatchSnapshot('warnings');

    await new Promise(async (resolve) => {
      const newStats = await compile(compiler);

      if (webpack.version[0] === '4') {
        expect(
          Object.keys(newStats.compilation.assets).filter(
            (assetName) => newStats.compilation.assets[assetName].emitted
          ).length
        ).toBe(7);
      } else {
        expect(newStats.compilation.emittedAssets.size).toBe(7);
      }

      expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
      expect(getWarnings(newStats)).toMatchSnapshot('errors');
      expect(getErrors(newStats)).toMatchSnapshot('warnings');

      resolve();
    });
  });
});
