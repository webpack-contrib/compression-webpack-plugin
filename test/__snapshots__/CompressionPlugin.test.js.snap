// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompressionPlugin should works (without options): assets 1`] = `
Array [
  "1.async.js?ver=1ea6f97e389bd951fd50",
  "5b1f36bc41ab31f5b801d48ba1d65781.png",
  "96621d3c37d96ad3bf792fcc848d912f.svg",
  "96621d3c37d96ad3bf792fcc848d912f.svg.gz",
  "js.js.gz?var=1ea6f97e389bd951fd50",
  "js.js?var=1ea6f97e389bd951fd50",
]
`;

exports[`CompressionPlugin should works (without options): errors 1`] = `Array []`;

exports[`CompressionPlugin should works (without options): warnings 1`] = `Array []`;
